src/sync/remote_mongo_collection.cpp

```diff
@@ -18,3 +18,505 @@
+{
+    update_many(filter_json, update_json, {}, completion_block);
+}
+
+void RemoteMongoCollection::find_one_and_update(const std::string& filter_json,
+                                                const std::string& update_json,
+                                                RemoteMongoCollection::RemoteFindOneAndModifyOptions options,
+                                                std::function<void(std::string, util::Optional<AppError>)> completion_block)
+{
+    try {
+        auto base_args = m_base_operation_args;
+        base_args.push_back({ "query", nlohmann::json::parse(filter_json) });
+        base_args.push_back({ "update", nlohmann::json::parse(update_json) });
+        //auto args = nlohmann::json( {{"arguments", nlohmann::json::array({base_args} ) }} );
+        
+        if (options.upsert) {
+            base_args.push_back({ "upsert", options.upsert });
+        }
+        
+        if (options.return_new_document) {
+            base_args.push_back({ "returnNewDocument", options.return_new_document });
+        }
+        
+        if (options.projection_json) {
+            base_args.push_back({ "project", nlohmann::json::parse(options.projection_json.value()) });
+        }
+        
+        if (options.projection_json) {
+            base_args.push_back({ "sort", nlohmann::json::parse(options.sort_json.value()) });
+        }
+        
+        auto args = nlohmann::json( {{"arguments", nlohmann::json::array({base_args} ) }} );
+
+        m_service->call_function("findOneAndUpdate",
+                                 args.dump(),
+                                 [completion_block](util::Optional<AppError> error, util::Optional<std::string> value) {
+            handle_response(error, value, completion_block);
+        });
+    } catch (const std::exception& e) {
+        return completion_block({}, AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+    }
+}
+
+void RemoteMongoCollection::find_one_and_update(const std::string& filter_json,
+                                                const std::string& update_json,
+                                                std::function<void(std::string, util::Optional<AppError>)> completion_block)
+{
+    find_one_and_update(filter_json, update_json, {}, completion_block);
+}
+
+void RemoteMongoCollection::find_one_and_replace(const std::string& filter_json,
+                                                 const std::string& replacement_json,
+                                                 RemoteMongoCollection::RemoteFindOneAndModifyOptions options,
+                                                 std::function<void(std::string, util::Optional<AppError>)> completion_block)
+{
+    try {
+        auto base_args = m_base_operation_args;
+        base_args.push_back({ "query", nlohmann::json::parse(filter_json) });
+        base_args.push_back({ "update", nlohmann::json::parse(replacement_json) });
+        
+        if (options.upsert) {
```

Comment:
consider moving this logic into `RemoteFindOneAndModifyOptions` (something like `std::vector<nlohmann::json> get_json()`) because that would clean this up and allow you to reuse the logic in several places